### PR TITLE
Reimplement pg_ctl start in python to simplify port opening detection

### PR DESCRIPTION
**Pull request by CyberDem0n**

### CODE REVIEW

File: patroni/postgresql.py [hunk 1]
```diff
@@ -698,31 +690,28 @@ def start(self, block_callbacks=False):
...
-        env = {'PATH': os.environ.get('PATH')}
-        # pg_ctl will write a FATAL if the username is incorrect. exporting PGUSER if necessary
-        if 'username' in self._superuser and self._superuser['username'] != os.environ.get('USER'):
-            env['PGUSER'] = self._superuser['username']
-
         self._write_postgresql_conf()
         self.resolve_connection_addresses()
 
-        options = ' '.join("--{0}='{1}'".format(p, self._server_parameters[p]) for p, v in self.CMDLINE_OPTIONS.items()
+        options = ' '.join('--{0}="{1}"'.format(p, self._server_parameters[p]) for p, v in self.CMDLINE_OPTIONS.items()
```
Comment (ants) [inline patroni/postgresql.py]: I think server parameter values need to be escaped for shell here. Are there any sanity checks for server parameter keys in Patroni? If not, that might also have some security implications. Having write permissions for Patroni configuration will probably allow for arbitrary code execution no matter what we do, but it shouldn't be this easy :)

File: patroni/postgresql.py [hunk 2]
```diff
@@ -698,31 +690,28 @@ def start(self, block_callbacks=False):
...
-            # We want postmaster to open ports before we continue
-            try:
-                start_timeout = float(self.config.get('pg_ctl_timeout', 60))
-            except ValueError:
-                start_timeout = 60
-            ret = self.wait_for_port_open(start_timeout, start_initiated, start_completed)
+        pid = os.fork()
+        if pid == 0:
+            os.setsid()
+            os.execle("/bin/sh", "/bin/sh", "-c", cmd, {'PATH': os.environ.get('PATH')})
```
Comment (ants) [inline patroni/postgresql.py]: Not a Windows expert by far, but I'll hazard a guess that this doesn't work on Windows :) Do we care?

Before execing there is probably some cleanup stuff to do, I know that watchdog support wants that the watchdog device file handle is closed and not passed on to postgres. Probably good to close the rest too. Why did you decide to go through `/bin/sh` and `os.execle`? Using `subprocess.Popen` to execute postgres directly might be more portable and give us stuff like close_fds.

File: patroni/postgresql.py (hunk 1 above)
Comment (CyberDem0n) [reply patroni/postgresql.py]: https://github.com/zalando/patroni/blob/master/patroni/postgresql.py#L58

Only parameters defined in `CMDLINE_OPTIONS` will be passed via command line.
Also it contains function for checking values of parameters. So it should be safe.

Except cluster_name, listen_addresses and port...

File: patroni/postgresql.py (hunk 2 above)
Comment (CyberDem0n) [reply patroni/postgresql.py]: Yeah it wont work on windows. Actually I took this example from original pg_ctl.c: https://github.com/postgres/postgres/blob/master/src/bin/pg_ctl/pg_ctl.c#L472

Will try to rewrite it with Popen...

**Comment by CyberDem0n** [comment]
Addressed code review. Any other comments?
Actually if you really want avoid  watchdog file handle being inherited by child processes and be always safe you should set `FD_CLOEXEC` flag on it,  we already doing so for REST API file handle: https://github.com/ants/patroni/blob/master/patroni/api.py#L427

**Comment by ants** [comment]
Checking it out now, had to get a couple of things off my plate first. Good idea with `FD_CLOEXEC`, that addresses my main concern.

**Pull request merged**